extensions/amp-story/1.0/amp-story-background-blur.js

```diff
@@ -0,0 +1,107 @@
+/**
+ * Copyright 2021 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {setStyles} from '#core/dom/style';
+
+/** @const {number} */
+const CANVAS_SIZE = 3;
+
+/** @const {number} */
+const DURATION_MS = 200;
+
+export class AmpStoryBackgroundBlur {
+  /**
+   * @param {!Window} win
+   * @param {!Element} element
+   */
+  constructor(win, element) {
+    /** @private @const {!Element} */
+    this.element_ = element;
+
+    /** @private @const {!Window} */
+    this.win_ = win;
+
+    /** @private @const {!Element} */
+    this.canvas_ = null;
+
+    /** @private @const {!Element} */
+    this.context_ = null;
+  }
+
+  /**
+   * Setup canvas, context and attach it to the document.
+   */
+  attach() {
+    this.canvas_ = document.createElement('canvas');
+    this.context_ = this.canvas_.getContext('2d');
+    this.canvas_.width = this.canvas_.height = CANVAS_SIZE;
+    setStyles(this.canvas_, {
+      width: '100%',
+      height: '100%',
+      position: 'absolute',
+      left: 0,
+      top: 0,
+    });
+    this.element_.appendChild(this.canvas_);
+  }
+
+  /**
+   * Update the background to the specified page's background.
+   * @param {!Element} pageElement
+   */
+  update(pageElement) {
+    const fillElement = this.getBackgroundElement_(pageElement);
+    this.animate_(fillElement);
+  }
+
+  /**
+   * Animated background transition.
+   * @private
+   * @param {?Element} fillElement
+   */
+  animate_(fillElement) {
+    let startTime;
+    const nextFrame = (currTime) => {
+      if (!startTime) {
+        startTime = currTime;
+      }
+      const elapsed = currTime - startTime;
+      if (elapsed < DURATION_MS) {
+        const easing = 1 - Math.pow(1 - elapsed / DURATION_MS, 2);
+        this.context_.globalAlpha = easing;
+        if (fillElement) {
+          this.context_.drawImage(fillElement, 0, 0, CANVAS_SIZE, CANVAS_SIZE);
+        } else {
+          this.context_.fillRect(0, 0, CANVAS_SIZE, CANVAS_SIZE);
+        }
+        requestAnimationFrame(nextFrame);
```

Comment:
This has potential to kick off a new animation loop if someone is navigating quickly. 
It needs to cancel if a new loop is started.

Reply:
+1
let's do this in this PR? If you want an immediate followup that's good too, but let's get this one right